Modernize the RootChild component (try 2)

jsnajdr opened the pull request:
The first attempt (#32749) had to be reverted because it broke the `LanguagePicker` UI. Any click in the language picker modal would close the modal immediately.

Why? The JSX has the following structure:
```jsx
<button onClick={ this.toggleOpen }>
  <span>Change Language</span>
  { this.state.open && <Dialog /> }
</button>
```

Click events inside the `Dialog` bubble up to the `button` element, even though the button is not the dialog's DOM parent. But it's its React tree parent and React [bubbles the events synthetically](https://reactjs.org/docs/portals.html#event-bubbling-through-portals). Therefore, any click inside the dialog triggers the `toggleOpen` click handler on the button.

This didn't happen when `RootChild` mounted an independent React root and put the `RootChild` children into it.

I fixed the `LanguagePicker` markup, but we'll need to revisit all `RootChild` usages and check for similar event "leaks". Until that's done, this PR is a WIP draft with uncertain future.

jsnajdr committed:
Implement RootChild using React portal

jsnajdr committed:
Dialog: remove RootChild

blowery commented:
Files that use either root-child or dialog:

- [ ] client/blocks/post-share/sharing-preview-modal.jsx:1
- [ ] client/blocks/post-share/publicize-actions-list.jsx:1
- [ ] client/blocks/site-address-changer/dialog.jsx:1
- [ ] client/blocks/disconnect-jetpack/dialog.jsx:1
- [ ] client/blocks/inline-help/dialog.jsx:1
- [ ] client/blocks/privacy-policy-banner/privacy-policy-dialog.jsx:1
- [ ] client/blocks/term-form-dialog/index.jsx:1
- [ ] client/blocks/taxonomy-manager/list-item.jsx:1
- [ ] client/blocks/upload-image/index.jsx:1
- [ ] client/blocks/support-article-dialog/index.jsx:1
- [ ] client/blocks/edit-gravatar/index.jsx:1
- [ ] client/layout/community-translator/launcher.jsx:1
- [ ] client/layout/nps-survey-notice/index.jsx:1
- [ ] client/extensions/woocommerce/app/order/order-payment/dialog.js:1
- [ ] client/extensions/woocommerce/app/order/order-details/fee-dialog.js:1
- [ ] client/extensions/woocommerce/app/order/order-details/product-dialog.js:1
- [ ] client/extensions/woocommerce/app/order/order-fulfillment/index.js:1
- [ ] client/extensions/woocommerce/app/order/order-customer/dialog.js:1
- [ ] client/extensions/woocommerce/app/settings/payments/payment-method-bacs.js:1
- [ ] client/extensions/woocommerce/app/settings/payments/payment-method-cheque.js:1
- [ ] client/extensions/woocommerce/app/settings/payments/payment-method-edit-dialog.js:1
- [ ] client/extensions/woocommerce/app/settings/payments/payment-method-paypal.js:1
- [ ] client/extensions/woocommerce/app/settings/payments/stripe/payment-method-stripe-connected-dialog.js:1
- [ ] client/extensions/woocommerce/app/settings/payments/stripe/payment-method-stripe-setup-dialog.js:1
- [ ] client/extensions/woocommerce/app/settings/payments/stripe/payment-method-stripe-placeholder-dialog.js:1
- [ ] client/extensions/woocommerce/app/settings/payments/stripe/payment-method-stripe-key-based-dialog.js:1
- [ ] client/extensions/woocommerce/app/settings/payments/stripe/payment-method-stripe-complete-oauth-dialog.js:1
- [ ] client/extensions/woocommerce/app/settings/shipping/shipping-zone/shipping-zone-method-dialog.js:1
- [ ] client/extensions/woocommerce/app/settings/shipping/shipping-zone/shipping-zone-location-dialog.js:1
- [ ] client/extensions/woocommerce/app/settings/email/mailchimp/setup-mailchimp.js:1
- [ ] client/extensions/woocommerce/app/products/product-form-variations-table.js:1
- [ ] client/extensions/woocommerce/woocommerce-services/views/label-settings/add-credit-card-modal.js:1
- [ ] client/extensions/woocommerce/woocommerce-services/views/packages/package-dialog.js:1
- [ ] client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/packages-step/move-item.js:1
- [ ] client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/packages-step/add-item.js:1
- [ ] client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/index.js:1
- [ ] client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-refund-modal.js:1
- [ ] client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-reprint-modal.js:1
- [ ] client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-details-modal.js:1
- [ ] client/extensions/woocommerce/components/store-address/index.js:1
- [ ] client/extensions/zoninator/components/settings/zone/delete-zone-dialog.jsx:1
- [ ] client/my-sites/comments/comment/comment-html-editor/add-image-dialog.jsx:1
- [ ] client/my-sites/comments/comment/comment-html-editor/add-link-dialog.jsx:1
- [ ] client/my-sites/domains/domain-management/transfer/transfer-to-other-user/index.jsx:1
- [ ] client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx:1
- [ ] client/my-sites/domains/domain-management/transfer/transfer-to-other-site/confirmation-dialog.jsx:1
- [ ] client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx:1
- [ ] client/my-sites/domains/domain-management/dns/delete-email-forwards-dialog.jsx:1
- [ ] client/my-sites/domains/components/domain-warnings/pending-gsuite-tos-notice-dialog.jsx:1
- [ ] client/my-sites/checkout/checkout-thank-you/google-voucher/index.jsx:1
- [ ] client/my-sites/people/people-invites/index.jsx:1
- [ ] client/my-sites/marketing/connections/account-dialog.jsx:1
- [ ] client/my-sites/site-settings/delete-site-warning-dialog/index.jsx:1
- [ ] client/my-sites/site-settings/delete-site/index.jsx:1
- [ ] client/my-sites/site-settings/theme-setup-dialog/index.jsx:1
- [ ] client/my-sites/themes/thanks-modal.jsx:1
- [ ] client/components/email-verification/email-verification-dialog/index.jsx:1
- [ ] client/components/tinymce/plugins/simple-payments/dialog/index.jsx:1
- [ ] client/components/tinymce/plugins/simple-payments/dialog/memberships.jsx:1
- [ ] client/components/tinymce/plugins/wplink/dialog.jsx:1
- [ ] client/components/tinymce/plugins/embed/dialog.jsx:1
- [ ] client/components/tinymce/plugins/embed/test/dialog.js:1
- [ ] client/components/tinymce/plugins/calypso-alert/alert.jsx:1
- [ ] client/components/tinymce/plugins/contact-form/dialog/index.jsx:1
- [ ] client/components/tinymce/plugins/wpcom-charmap/charmap.jsx:1
- [ ] client/components/tinymce/plugins/wpcom-help/help-modal.jsx:1
- [ ] client/components/site-selector-modal/index.jsx:1
- [ ] client/components/marketing-survey/gsuite-cancel-purchase-dialog/index.jsx:1
- [ ] client/components/community-translator/translatable.jsx:1
- [ ] client/components/language-picker/modal.jsx:1
- [ ] client/me/notification-settings/push-notification-settings/index.jsx:1
- [ ] client/me/purchases/remove-purchase/index.jsx:1
- [ ] client/me/purchases/cancel-purchase/button.jsx:1
- [ ] client/me/reauth-required/index.jsx:1
- [ ] client/me/account-close/confirm-dialog.jsx:1
- [ ] client/post-editor/media-modal/dialog.jsx:1
- [ ] client/post-editor/editor-media-advanced/index.jsx:1
- [ ] client/post-editor/editor-more-options/copy-post.jsx:1
- [ ] client/post-editor/restore-post-dialog.jsx:1
- [ ] client/post-editor/editor-post-type-unsupported/index.jsx:1
- [ ] client/post-editor/editor-gutenberg-opt-in-dialog/index.jsx:1
- [ ] client/post-editor/invalid-url-dialog.jsx:1
- [ ] client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx:1
- [ ] client/post-editor/editor-revisions/dialog.jsx:1
- [ ] client/post-editor/editor-html-toolbar/add-image-dialog.jsx:1
- [ ] client/post-editor/editor-html-toolbar/add-link-dialog.jsx:1
- [ ] client/post-editor/editor-forbidden/index.jsx:1
- [ ] client/lib/accept/dialog.jsx:1
- [ ] client/lib/keyboard-shortcuts/menu.jsx:1
- [ ] client/jetpack-connect/sso.jsx:1

blowery committed:
Fixup nested dialog in button for jetpack disconnect

jsnajdr commented:
> Files that use either root-child or dialog:

Hmm, not terribly encouraging 😆 

The ESLint errors in `InfoPopover` deserve a separate PR: #32922

sgomes commented:
@jsnajdr : What's the current status on this PR? Do all of those checkboxes still need ticking, and if so would you like help doing it?

jsnajdr commented:
> What's the current status on this PR? Do all of those checkboxes still need ticking, and if so would you like help doing it?

@sgomes Yes, these checkboxes need ticking, and help is appreciated! What we need to check for is that we don't embed a `RootChild`-using component (`Popover`, `Dialog`) inside a `Button` or other clickable element:
```js
<Button onClick={ handleClick }>
  Open Dialog
  <Dialog isVisible={ showDialog } />
</Button>
```
With React Portal, clicks inside the `Dialog` will bubble up to the `Button` and will incorrectly invoke the `handleClick` handler. We need to refactor that to this:
```js
<Button onClick={ handleClick }>Open Dialog</Button>
<Dialog isVisible={ showDialog } />
```

Good first step is to review #36561 (spinoff from this PR) and to get the idea 🙂

sgomes commented:
> > What's the current status on this PR? Do all of those checkboxes still need ticking, and if so would you like help doing it?
> 
> @sgomes Yes, these checkboxes need ticking, and help is appreciated! What we need to check for is that we don't embed a `RootChild`-using component (`Popover`, `Dialog`) inside a `Button` or other clickable element:
> 
> ```js
> <Button onClick={ handleClick }>
>   Open Dialog
>   <Dialog isVisible={ showDialog } />
> </Button>
> ```
> 
> With React Portal, clicks inside the `Dialog` will bubble up to the `Button` and will incorrectly invoke the `handleClick` handler. We need to refactor that to this:
> 
> ```js
> <Button onClick={ handleClick }>Open Dialog</Button>
> <Dialog isVisible={ showDialog } />
> ```
> 
> Good first step is to review #36561 (spinoff from this PR) and to get the idea 🙂

Thanks for the explanation, @jsnajdr! I'll be reviewing some of these components and, where applicable, the code making use of them 👍

sgomes commented on client/blocks/edit-gravatar/index.jsx:
```diff
@@ -215,6 +214,7 @@ export class EditGravatar extends Component {
 						</div>
 					</FilePicker>
 				</div>
+				{ this.renderImageEditor() }
```
FYI, I have #36656 out with this change already.

sgomes replied:
And it's now merged, thanks to your review, so a rebase will pull it in 👍

jsnajdr replied:
Rebased, and the commit magically disappeared :)

jsnajdr committed:
Fixup the Themes SSR tests: no DOM, alias Popover component

SSR tests shouldn't run in a JSDOM environment, as that prevents bugs that
reference `window` or `document` from failing the tests.

Also, alias the `Popover` component to empty, so that the `click-outside` and
`RootChild` modules that need DOM don't trigger errors. The webpack config
that builds the server bundle does that aliasing, too.

jsnajdr committed:
AccountDialog: reenable the jsx-classname-namespace ESLint rule

jsnajdr committed:
DeleteSiteWarningDialog: reactive useTranslate hook, use Button component

jsnajdr committed:
EditGravatar: move email verification dialog out of clickable area

jsnajdr committed:
Masterbar Publish Button: move SitesPopover out of clickable Button el

jsnajdr committed:
Accordion Status: move tooltip out of clickable area

jsnajdr committed:
Post Editor: move Tooltips outside hover areas

jsnajdr committed:
Move more Tooltips out of clickable/hoverable areas

jsnajdr commented:
OK, this is finally ready for review and merging. I went through all usages of `<Dialog>`, `<Popover>` and `<Tooltip>` and fixed them where necessary -- moving them out of elements with click or hover event handlers.

I also had to fixup tests for Themes SSR: in SSR, React Portals shouldn't render and the components that use them (i.e., modals and popups) should render as `null`. I achieved that by aliasing `components/popover` to an empty package in the test. The webpack config for the server build already does that aliasing.

Some things I fixed:

Accordion status icon in Classic Editor:
<img width="275" alt="Screenshot 2019-11-04 at 10 35 44" src="https://user-images.githubusercontent.com/664258/68124550-f619f780-ff0f-11e9-8fcc-4859c925f7eb.png">

Tooltips for Classic Editor UI buttons and many other tooltips:
<img width="241" alt="Screenshot 2019-11-04 at 14 34 01" src="https://user-images.githubusercontent.com/664258/68124683-337e8500-ff10-11e9-83e8-5136412b70ee.png">
<img width="303" alt="Screenshot 2019-11-04 at 14 34 15" src="https://user-images.githubusercontent.com/664258/68124688-37aaa280-ff10-11e9-8d31-626d20f1561f.png">

Publish button in Masterbar -- it would close the `SitesPopover` when clicking inside the popover, because the click would bubble up to the button and trigger the toggling handler:
<img width="345" alt="Screenshot 2019-11-04 at 14 35 26" src="https://user-images.githubusercontent.com/664258/68124800-78a2b700-ff10-11e9-8cc4-3ae8b799ef5c.png">

This PR is a major step to moving a lot of components (`Dialog`, `Popover`, `Tooltip`, ...) to `@automattic/components`, because if frees the `RootChild` from depending on app-specific providers like Redux.

jsnajdr committed:
AccordionStatus: update unit tests to work with the new nested Fragment

sgomes commented on client/components/root-child/index.jsx:
```diff
@@ -1,67 +1,21 @@
-
 	render() {
-		return null;
+		return ReactDOM.createPortal( this.props.children, this.container );
```
\o/

sgomes commented on client/my-sites/comments/comment/comment-author.jsx:
```diff
@@ -34,10 +34,7 @@ export class CommentAuthor extends Component {
 		isLinkTooltipVisible: false,
 	};
 
-	shouldComponentUpdate = ( nextProps, nextState ) =>
```
Good call. This attempted optimisation seems like overkill for what's mostly a leaf component.

jsnajdr replied:
Yes! It also used the deep `isEqual` comparison, which is a red flag.

sgomes commented on client/my-sites/plugins/disconnect-jetpack/disconnect-jetpack-button.jsx:
```diff
@@ -57,29 +57,30 @@ class DisconnectJetpackButton extends Component {
-					translate( 'Disconnect', {
-						context: 'Jetpack: Action user takes to disconnect Jetpack site from .com',
-					} ) }
+			<>
```
Nit: Suggest using `<Fragment>` here, for consistency with the rest of the PR.

jsnajdr replied:
Done in https://github.com/Automattic/wp-calypso/pull/32899/commits/dc2b31136b76d193d3106e69007c176abfd4a27c

sgomes commented on client/my-sites/site-settings/delete-site-warning-dialog/index.jsx:
```diff
@@ -1,46 +1,48 @@
-		</p>
-	</Dialog>
-);
+const DeleteSiteWarningDialog = ( { isVisible, onClose } ) => {
```
Nit: make this a function instead of an arrow function.

jsnajdr replied:
Done in https://github.com/Automattic/wp-calypso/pull/32899/commits/dcdcccaa1f7cb4b220d0b623ce8229dc4b71f73d

sgomes reviewed (approved):
LGTM, code-wise. I'll try to run some tests once I work out some login issues I'm having in my browser.

sgomes commented:
FYI, I tested a few of the updated components, and they seem to be working nicely 👍

jsnajdr committed:
DeleteSiteWarningDialog: turn into arrowless function

jsnajdr committed:
DisconnectJetpackButton: use Fragment

jsnajdr commented:
Thanks @sgomes for review! I'll check the e2e failures re Simple Payments Dialog and then merge.

jsnajdr committed:
Rewrite RootChild to stop rendering into a detached DOM container

Avoids some subtle bugs with the rendered DOM not being measurable
or `iframe.contentWindow` being `null`.

jsnajdr commented:
e2e tests started reporting failures when testing post preview and trying to reproduce I confirmed that they were right ☹️ 

`RootChild` had a very subtle bug: it created a container DOM element, then rendered the portal into it, and only a moment later, in `componentDidMount`, did it add the container DOM element to the `document.body`.

That means that the `RootChild` children were rendered into a 'detached' DOM element, and when the children components' lifecycles and hooks tried to access the DOM, they were accessing a detached DOM. That DOM is not part of the document and its layout though, and therefore doesn't have dimensions. Components that rightfully depend on layout being finished can fail in subtle ways.

Another thing that doesn't work in detached DOM is `iframe.contentWindow`. The `iframe` is not yet 'active', and its `contentWindow` is `null`. The HTML spec has some [incomprehensible language about nested browsing contexts](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#dom-iframe-contentwindow) if you're interested in details 🙂 

`iframe.contentWindow` is exactly what [WebPreview relies on](https://github.com/Automattic/wp-calypso/blob/0b2f5996067a62ef8705d148900cf3d6a06fc08d/client/components/web-preview/content.jsx#L184) when setting the `iframe`'s preview URL. It does something like:
```js
class WebPreviewContent extends Component {
  componentDidMount() {
    this.iframeRef.current.contentWindow.location.replace( ... );
  }

  render() {
    return <iframe ref={ this.iframeRef } />;
  }
}

function WebPreviewModal() {
  return <RootChild><WebPreviewContent /></RootChild>;
}
```
The `contentWindow.location` throws a reference error when running on detached DOM.

The original implementation of `RootChild` was inspired by `react-modal`, which also has this bug. I rewrote it in a way used by [Material UI's `Portal`](https://github.com/mui-org/material-ui/blob/master/packages/material-ui/src/Portal/Portal.js).

jsnajdr commented:
All e2e tests are instantly green! Let's merge and watch the production closely today.
